corehq/elastic.py

```diff
@@ -289,6 +289,49 @@ def scan(client, query=None, scroll='5m', **kwargs):
         )
 
     """
+    if settings.ELASTICSEARCH_MAJOR_VERSION == 7:
+        return scan_es7(client, query=query, scroll=scroll, **kwargs)
+    else:
+        return scan_es2(client, query=query, scroll=scroll, **kwargs)
```

Comment:
What are the actual differences between these two versions?  Is the ES7 version for sure not available on ES2 as well?

It looks like `scan` is only called in one place anyways (`scroll_query`, just above this), and that is only called in one place (`ESQuery.scroll`), so I think there's actually a lot less variation than this stuff suggests.  In light of that, I'm betting this can be simplified significantly.  I see the docstring says

> This is a copy of elasticsearch.helpers.scan, except this function returns a ScanResult (which includes the total number of documents), and removes some options from scan that we aren't using.

I bet we can actually drop both of these functions and use that [`scan` library](https://elasticsearch-py.readthedocs.io/en/master/helpers.html?#scan) directly.  I think the only reason we reach in to the internals here is to get a count of documents without doing a second query, but looking at the places `.scroll()` is called, I don't actually see anywhere that uses that functionality - they all just treat it as an iterable, so that point is moot.  [These](https://github.com/dimagi/commcare-hq/blob/13228320848a504aa5e1b3d0488998e22e3e6c9d/corehq/apps/dump_reload/management/commands/dump_domain_data_raw.py#L89-L94) [two](https://github.com/dimagi/commcare-hq/blob/13228320848a504aa5e1b3d0488998e22e3e6c9d/corehq/apps/reports/standard/deployments.py#L434-L436) instances _could_ actually use that functionality, but I think I actually prefer a separate `count` query.

I think the possible approaches here are:

* Keep mostly as-is - I'd then simplify the code a bit so it's easier to follow, and try to unify as much as possible between the two versions if they really can't be unified
* Drop this entirely and make `ESQuery.scroll` call `elasticsearch.helpers.scan` and return a simple generator rather than a `ScanResult`